frontend/src/App.tsx

```diff
@@ -841,6 +841,23 @@ export class App extends PureComponent<Props, State> {
       )
       this.widgetMgr.removeInactive(activeWidgetIds)
 
+      // Tell the ConnectionManager to increment the message cache run
+      // count. This will result in expired ForwardMsgs being removed from
+      // the cache.
+      if (this.connectionManager !== null) {
+        this.connectionManager.incrementMessageCacheRunCount(
+          SessionInfo.current.maxCachedMessageAge
+        )
+      }
+    } else if (
+      status === ForwardMsg.ScriptFinishedStatus.FINISHED_EARLY_FOR_RERUN
+    ) {
+      // Notify any subscribers of this event (and do it on the next cycle of
+      // the event loop)
+      window.setTimeout(() => {
+        this.state.scriptFinishedHandlers.map(handler => handler())
+      }, 0)
```

Comment:
Oh, I think I see - it looks like this is logic duplicated from above. My previous comments still stand, but I wonder if it's also worth de-duping this logic - pull it out of these two `if` clauses, and move it to the bottom of the function in an `if status == FINISHED_SUCCESSFULLY || status == FINISHED_EARLY` clause?

Same with the "increment the message cache run count" logic. (And then the related tests - if we have any, and it looks like we may not, which is a bit of a yikes - should be updated to include these new conditions.)

Reply:
I have deduped the logic, but I don't know why the handlers need to happen on the next tick, since it was already like that, so I can't expand on the comment